share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java

```diff
@@ -0,0 +1,136 @@
+        minOffset.set(Long.MAX_VALUE);  // For easy min update.
+        redundantOffset = new TimelineLong(snapshotRegistry);
+    }
+
+    /**
+     * Method updates internal state with the supplied offset for the provided
+     * share partition key. It then calculates the minimum offset, if possible,
+     * below which all offsets are redundant. This value is then returned as
+     * an optional.
+     * <p>
+     * The value returned is exclusive, in that all offsets below it but not including
+     * it are redundant.
+     *
+     * @param key    - represents {@link SharePartitionKey} whose offset needs updating
+     * @param offset - represents the latest partition offset for provided key
+     */
+    public void updateState(SharePartitionKey key, long offset) {
+        minOffset.set(Math.min(minOffset.get(), offset));
+        offsets.put(key, offset);
+
+        Optional<Long> deleteTillOffset = findRedundantOffset();
+        deleteTillOffset.ifPresent(off -> {
+            minOffset.set(off);
+            redundantOffset.set(off);
+        });
+    }
+
+    private Optional<Long> findRedundantOffset() {
+        if (offsets.isEmpty()) {
+            return Optional.empty();
+        }
+
+        long soFar = Long.MAX_VALUE;
+
+        for (long offset : offsets.values()) {
+            // Get min offset among latest offsets
+            // for all share keys in the internal partition.
+            soFar = Math.min(soFar, offset);
+
+            // minOffset represents the smallest necessary offset
+            // and if soFar equals it, we cannot proceed. This can happen
+            // if a share partition key hasn't had records written for a while
+            // For example,
+            // <p>
+            // key1:1
+            // key2:2 4 6
+            // key3:3 5 7
+            // <p>
+            // We can see in above that offsets 2, 4, 3, 5 are redundant,
+            // but we do not have a contiguous prefix starting at minOffset
+            // and we cannot proceed.
+            if (soFar == minOffset.get()) {
+                return Optional.empty();
+            }
+        }
+
+        return Optional.of(soFar);
+    }
+
+    /**
+     * Most recent last redundant offset. This method is to be used
+     * when the caller wants to query the value of such offset.
+     * @return Optional of type Long representing the offset or empty for invalid offset values
+     */
+    public Optional<Long> lastRedundantOffset() {
+        long value = redundantOffset.get();
+        if (value <= 0 || value == Long.MAX_VALUE) {
+            return Optional.empty();
+        }
+        // We don't want to send the same value repeatedly
+        // as caller might call delete records again and again
+        // reducing efficiency hence, once valid redundant value is
+        // returned lets set it to 0 so next time onwards
+        // empty Optional is returned.
+        redundantOffset.set(0L);
```

Comment:
Hmm, should we confirm that the subsequent `deleteRecord` call is successful before resetting `redundantOffset`?

Reply:
That is not the responsibility of this class. 

The call to this method is from a periodic timer task which keeps querying the method and if a valid value is found - issue the `ReplicaManager.deleteRecords` call. Since the redundant offset will monotonically increase, even if the `deleteRecord` fails for some reason for a lower offset - a higher redundant offset will result in deletion later on.

This value is being set to 0 here to prevent repeated calls to `ReplicaManager.deleteRecords` which will not be made if an empty optional (as a consequence of redundant offset being 0) is returned.

Reply:
@junrao